granulate_utils/linux/elf.py

```diff
@@ -90,3 +97,52 @@ def is_statically_linked(path: str) -> bool:
             if segment.header.p_type == "PT_DYNAMIC":
                 return False
     return True
+
+
+def get_symbol_addr(path: str, sym_name: str) -> Optional[int]:
+    with open(path, "rb") as f:
+        elf = ELFFile(f)
+        symtab = elf.get_section_by_name(".symtab")
+        if symtab is None:
+            return None
+        symbols = symtab.get_symbol_by_name(sym_name)
+        if symbols is None or len(symbols) != 1:
+            return None
+        return symbols[0].entry.st_value
+
+
+def get_dt_needed(path: str) -> List[str]:
+    dt_needed: List[str] = []
+    with open(path, "rb") as f:
+        elf = ELFFile(f)
+        dynamic_section = elf.get_section_by_name(".dynamic")
+        if dynamic_section is None:
+            return dt_needed
+        for tag in dynamic_section.iter_tags():
+            if tag.entry.d_tag == "DT_NEEDED":
+                dt_needed.append(tag.needed)
+    return dt_needed
+
+
+def get_libc_type(path: str) -> LibcType:
+    dt_needed = get_dt_needed(path)
+    found_glibc = False
+    found_musl = False
+    for needed in dt_needed:
+        if "libc.so.6" in needed:
+            found_glibc = True
+        if "libc.musl" in needed:
+            found_musl = True
+        if "ld-musl" in needed:
+            found_musl = True
+    if found_glibc and found_musl:
+        raise Exception("Found both musl and glibc in the same binary")
+    if found_musl:
+        return LibcType.DYNAMIC_MUSL
+    if found_glibc:
+        return LibcType.DYNAMIC_GLIBC
+
+    if get_symbol_addr(path, "__libc_start_main") is not None:
```

Comment:
Is this true for both glibc and musl?

Reply:
Yep - musl even uses this to discover if you are trying to dlopen another libc to redirect you to dlopen libc.so, with a comment stating this should catch both glibc and musl, so I am pretty confident:

![image](https://user-images.githubusercontent.com/3260178/194286858-580682b6-0f9c-4092-90ed-aa0ed1a32b59.png)

Reply:
Cool trick. Document it here, permalink to this comment would do.

Also - seems like this logic checks for both syms, not just `__libc_start_main`, any reason you're not following it?

Reply:
Also, code checks for != 0 and you check for "existing", but I guess that's enough as well

Reply:
> Also - seems like this logic checks for both syms, not just __libc_start_main, any reason you're not following it?

I am using this image as a proof that "glibc and musl always have __libc_start_main", not as "this is the way to check".
They are using two symbols to decrease the chance for a false positive and loading the wrong thing. For use, we have a static binary that we just want to know if it uses a libc or if it is "bare" like staic go binaries, so just that symbol is enough to say "it has a libc", we would rather make the mistake in the other direction.

I can add a validation that `stdin` exists and raise an exception if it doesn't - if you feel it is safer, but returning NO_LIBC after finding `__libc_start_main` just because `stdin` is not there just sounds plain wrong.

Reply:
Adding a descriptive comment and link to musl code

Reply:
> I can add a validation that stdin exists and raise an exception if it doesn't - if you feel it is safer, but returning NO_LIBC after finding __libc_start_main just because stdin is not there just sounds plain wrong.

Sounds okay.

I am resolving this comment